Automattic/wp-calypso
# Migrate lib/mixins/infinite-scroll to a React component

While working on #23090, I stumbled upon a React component that still uses `createReactClass` because of the `infinite-scroll` mixin. I decided to migrate that mixin to a React component with exactly the same functionality.

There are a few step-by-step commits here that:
- add the new component,
- migrate the existing usages,
- remove the unused mixin
- codemod the old-style components to an ES6 class

Answers to some questions that a potential reviewer might ask:
**Why is it a component, rather than a HoC?**
It has to do with the unidirectional data flow. HoC is suitable for situations where the HoC supplies extra properties to the wrapped component. But in this case, the component that uses `InfiniteScroll` passes a callback as the `fetchNextPage` prop to it.

**Does the `InfiniteScroll`'s `componentDidUpdate` get called every time the parent component gets rerendered?**
Yes. When the parent component rerenders for any reasons, the `InfiniteScroll` is rerendered, too and its `componentDidUpdate` is called, although its render method returns `null` and the props haven't changed. The behavior is identical to the original mixin.

**How to test:**
There are three components affected:
- list of site pages (I used Domains University for testing, as it has a lot of pages)
- list of plugins when searching for some string (e.g., "SEO" returns a lot of results)
- the list of themes

Verify that all three load new items after scrolling to the bottom.
